### PR TITLE
feat(recipes): modernize DeepSeek V4 library configs

### PR DESCRIPTION
**Pull request by cuichenx**
## What does this PR do?

- modernizes the DeepSeek V4 GB200 library recipes while leaving `perf_recipes` unchanged
- adds a 128-GPU GB200 MXFP8 pretraining recipe with PP1/EP64, grouped GEMM, HybridEP, selective recompute, attention activation offload, and MXFP8 parameter gather/buffer reuse
- adds a GB200 offline-packed OpenMath SFT recipe with grouped GEMM, HybridEP, fixed packed-sequence shapes, DSA execution fusion, selective recompute, and attention activation offload
- keeps natural routing, unlimited expert capacity, full-precision gradient reduction, validation checks, optimizer semantics, and checkpoint behavior; it does not port forced balancing, paged stash, token dropping, reduced-precision gradients, or CUDA graphs into the library recipe
- applies recipe environment variables before training imports for both library and benchmark recipes
- preserves legacy DeepSeek V4 indexer scorer names during BF16 export
- adds focused recipe, runner, conversion, and verification-card coverage
- adds the DeepSeek V4 Flash verification card with the final pretraining and SFT evidence

DeepSeek V4 pretraining requires Megatron-LM `dev`. Grouped-FP8 checkpoint saves also require the empty extra-state fix from NVIDIA/Megatron-LM#5997. Until that fix reaches `dev`, use the equivalent signed commit from NVIDIA/Megatron-LM#6509 on a compatible `dev` checkout.

## Validation

| Workload | Configuration | Result | Last-10-step average |
| --- | --- | --- | ---: |
| Pretrain | 128 GB200, MXFP8, PP1/EP64, natural routing | 100 finite steps; grouped-MXFP8 checkpoint at step 100; fresh model/optimizer/RNG restore completed steps 101-105 | 131.36 TFLOP/s/GPU, 5713.47 ms |
| Packed SFT | 32 GB200, BF16, PP4/EP8 | 100 finite steps; fresh checkpoint reload completed step 101 | 42.18 TFLOP/s/GPU, 8575.85 ms |

Both runs completed with zero skipped or NaN iterations. The pretraining loss moved from `12.59201` to `5.960634`; packed-SFT loss moved from `0.9435847` to `0.2834835`.

Additional checks:

- focused branch test matrix: `298 passed, 1 skipped`
- latest-`main` integrated focused matrix: `335 passed, 1 skipped, 1 deselected`
- grouped checkpoint tests: `15 passed`
- all-file pre-commit: passed
- verification-card validator: passed on the latest-`main` integrated branch
- `git diff --check` and public-diff privacy scan: passed

The deselected metadata test has an unrelated Nemotron 3 Super recipe-name collision that reproduces on current `main`; this PR does not modify that metadata or either Nemotron recipe export.

## Verification scope

The card marks GB200 pretraining and packed SFT verified. Formal checkpoint-resume comparison remains unverified because its contract requires an independent step-50-to-100 run with step-51 and step-100 sentinels. Public post-SFT export/inference, long-context SFT, and the canonical mock-data performance item also remain explicitly unverified.

A capacity-1.5 performance screen was rejected for the library recipe because it dropped more than 64% of expert assignments and routing probability mass. The verified library recipe therefore retains unlimited-capacity natural routing.

- [x] I, the PR author, have personally reviewed every line of this PR.

### CODE REVIEW

File: scripts/training/run_recipe.py [hunk 1]
```diff
@@ -106,7 +106,6 @@
 from recipe_runner import (  # noqa: E402
     apply_cli_overrides,
     apply_determinism,
-    apply_runtime_environment,
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py:109]: why this file need to change? it seems will change the behaviors

Comment (cuichenx) [reply scripts/training/run_recipe.py:109]: good catch i'll make this more narrow

**Review by yaoyu-33 — DISMISSED** [review]

File: src/megatron/bridge/recipes/deepseek/gb200/deepseek_v4.py [hunk 2]
```diff
@@ -189,6 +191,66 @@ def deepseek_v4_flash_pretrain_64gpu_gb200_fp8mx_config() -> ConfigContainer:
     return cfg
 
 
+def deepseek_v4_flash_pretrain_128gpu_gb200_fp8mx_library_config() -> ConfigContainer:
+    """Return the scalable real-training DeepSeek V4 Flash GB200 config.
+
+    The historical name is retained for compatibility. This variant uses PP4/VPP4,
+    EP16, and selective recompute without activation offload; its current
+    qualification is at 64 GPUs. Expert capacity, paged stash, CUDA graphs, and
```
Comment (weijiac0619) [inline src/megatron/bridge/recipes/deepseek/gb200/deepseek_v4.py]: the name is for 128 gpus?

File: src/megatron/bridge/recipes/deepseek/gb200/deepseek_v4.py [hunk 3]
```diff
@@ -189,6 +191,66 @@ def deepseek_v4_flash_pretrain_64gpu_gb200_fp8mx_config() -> ConfigContainer:
...
+    cfg.model.moe_token_dispatcher_type = "flex"
+    cfg.model.moe_flex_dispatcher_backend = "hybridep"
+    cfg.model.moe_shared_expert_overlap = False
+    cfg.model.moe_flex_dispatcher_num_sms = 32
+    cfg.model.moe_hybridep_num_sms_preprocessing = 108
+    cfg.model.moe_mlp_glu_interleave_size = 32
+    cfg.model.use_transformer_engine_op_fuser = True
+    cfg.model.recompute_modules = ["mhc", "mla_up_proj"]
+    cfg.model.fine_grained_activation_offloading = False
+    cfg.model.offload_modules = []
```
Comment (weijiac0619) [inline src/megatron/bridge/recipes/deepseek/gb200/deepseek_v4.py]: nit: in peft it's set to none. equivalent to setting it to [] here since mcore handles that. do we need to keep it consistent?
